homeassistant/components/mcp/coordinator.py

```diff
@@ -0,0 +1,79 @@
+"""Types for the Model Context Protocol integration."""
+
+from collections.abc import AsyncGenerator
+from contextlib import AbstractAsyncContextManager, asynccontextmanager
+import datetime
+import logging
+
+import httpx
+from mcp.client.session import ClientSession
+from mcp.client.sse import sse_client
+from mcp.types import Tool
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_URL
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+UPDATE_INTERVAL = datetime.timedelta(minutes=30)
+
+
+@asynccontextmanager
+async def mcp_client(url: str) -> AsyncGenerator[ClientSession]:
+    """Create a server-sent event MCP client.
+
+    This is an asynccontxt manager that wraps to other async context managers
+    so that the coordinator has a single object to manage.
+    """
+    try:
+        async with sse_client(url=url) as streams, ClientSession(*streams) as session:
+            await session.initialize()
+            yield session
+    except ExceptionGroup as err:
+        raise err.exceptions[0] from err
+
+
+class ModelContextProtocolCoordinator(DataUpdateCoordinator[list[Tool]]):
+    """Define an object to hold MCP data."""
+
+    config_entry: ConfigEntry
+    session: ClientSession
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
```

Comment:
Let's use the typed config entry here

Reply:
I did not do that here because that adds a circular import because the config entry type has a runtime data that is the coordinator. (It seemed OK to use the raw config entry here because nothing in the coordinator should depend on the runtime data since its point to itself)

Is there a typical way integrations handle this?

Reply:
Usually define the config entry type here. I am also not 100% convinced, but the argument for doing that is added consistency

Reply:
Fixed. Agreed on consistency being the important factor. (Just read this yesterday https://www.seangoedecke.com/large-established-codebases/ on HN.)

Reply:
Sounds like I got something to read during dinner, thanks!